engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/PassManager.scala

```diff
@@ -198,27 +153,97 @@ class PassManager(
     val newContext =
       inlineContext.copy(passConfiguration = Some(passConfiguration))
 
-    val passesWithIndex = passGroup.passes.zipWithIndex
+    runPasses[Expression, InlineContext](
+      ir,
+      newContext,
+      passGroup,
+      createMiniPass =
+        (factory, ctx) => factory.createForInlineCompilation(ctx),
+      miniPassCompile = (miniPass, ir) =>
+        MiniIRPass.compile[Expression](classOf[Expression], ir, miniPass),
+      megaPassCompile = (megaPass, ir, ctx) => megaPass.runExpression(ir, ctx)
+    )
+  }
 
-    passesWithIndex.foldLeft(ir) {
-      case (intermediateIR, (pass, index)) => {
+  /** Runs all the passes in the given `passGroup` on `ir` with `context`.
+    * @param createMiniPass Function that creates a minipass.
+    * @param miniPassCompile Function that compiles IR with mini pass.
+    * @param megaPassCompile Function that compiles IR with mega pass.
+    * @tparam IRType Type of the [[IR]] that is being compiled.
+    * @tparam ContextType Type of the context for the compilation.
+    *                     Either [[ModuleContext]] or [[InlineContext]]
+    * @return Compiled IR. Might be the same reference as `ir` if no compilation was done.
+    */
+  private def runPasses[IRType <: IR, ContextType](
```

Comment:
One method to do the loop used from both `runPassesOnModule` and `runPassesInline` is a great improvement! I wanted it, but I got scared by the necessary typings when [combining subsequent minipasses](https://github.com/enso-org/enso/pull/11191/commits/2736a7642886c9a8919fb037ef21463422a2c001).